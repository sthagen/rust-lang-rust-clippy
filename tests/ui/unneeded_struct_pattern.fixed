//@aux-build:non-exhaustive-enum.rs
#![allow(
    clippy::manual_unwrap_or_default,
    clippy::manual_unwrap_or,
    clippy::redundant_pattern_matching,
    clippy::no_effect
)]
#![warn(clippy::unneeded_struct_pattern)]

extern crate non_exhaustive_enum;
use non_exhaustive_enum::*;

fn noop() {}

fn main() {
    match Some(114514) {
        Some(v) => v,
        None => 0,
        //~^ unneeded_struct_pattern
    };

    match Some(1919810) {
        Some(v) => v,
        None => 0,
        //~^ unneeded_struct_pattern
    };

    match Some(123456) {
        Some(v) => v,
        None => 0,
    };

    match Some(Some(123456)) {
        Some(Some(v)) => v,
        Some(None) => 0,
        //~^ unneeded_struct_pattern
        None => 0,
        //~^ unneeded_struct_pattern
    };

    if let None = Some(0) {}
    //~^ unneeded_struct_pattern
    if let None = Some(0) {}
    //~^ unneeded_struct_pattern
    if let Some(None) = Some(Some(0)) {}
    //~^ unneeded_struct_pattern
    let None = Some(0) else { panic!() };
    //~^ unneeded_struct_pattern
    let None = Some(0) else { panic!() };
    //~^ unneeded_struct_pattern
    let Some(None) = Some(Some(0)) else { panic!() };
    //~^ unneeded_struct_pattern

    enum Custom {
        HasFields {
            field: i32,
        },
        HasBracketsNoFields {},
        NoBrackets,
        #[non_exhaustive]
        NoBracketsNonExhaustive,
        Init,
    };

    match Custom::Init {
        Custom::HasFields { field: value } => value,
        Custom::HasBracketsNoFields {} => 0,
        _ => 0,
    };

    match Custom::Init {
        Custom::HasFields { field: value } => value,
        Custom::HasBracketsNoFields { .. } => 0,
        _ => 0,
    };

    0;
    //~^^^ match_single_binding

    0;

    if let Custom::HasFields { field: value } = Custom::Init {
        noop();
    }
    if let Custom::HasBracketsNoFields {} = Custom::Init {
        noop();
    }
    if let Custom::HasBracketsNoFields { .. } = Custom::Init {
        noop();
    }
    if let Custom::NoBrackets = Custom::Init {
        //~^ unneeded_struct_pattern
        noop();
    }
    if let Custom::NoBrackets = Custom::Init {
        //~^ unneeded_struct_pattern
        noop();
    }
    if let Custom::NoBrackets | Custom::NoBracketsNonExhaustive = Custom::Init {
        //~^ unneeded_struct_pattern
        //~| unneeded_struct_pattern
        noop();
    }
    if let Custom::NoBracketsNonExhaustive = Custom::Init {
        //~^ unneeded_struct_pattern
        noop();
    }
    if let Custom::NoBracketsNonExhaustive = Custom::Init {
        //~^ unneeded_struct_pattern
        noop();
    }

    let Custom::HasFields { field: value } = Custom::Init else {
        panic!()
    };

    let Custom::HasBracketsNoFields {} = Custom::Init else {
        panic!()
    };

    let Custom::HasBracketsNoFields { .. } = Custom::Init else {
        panic!()
    };

    let Custom::NoBrackets = Custom::Init else {
        //~^ unneeded_struct_pattern
        panic!()
    };
    let Custom::NoBracketsNonExhaustive = Custom::Init else {
        //~^ unneeded_struct_pattern
        panic!()
    };
    let Custom::NoBracketsNonExhaustive = Custom::Init else {
        //~^ unneeded_struct_pattern
        panic!()
    };

    enum Refutable {
        Variant,
    }
}

fn external_crate() {
    use ExtNonExhaustiveVariant::*;

    match ExhaustiveUnit {
        // Expected
        ExhaustiveUnit => 0,
        _ => 0,
    };

    0;

    0;

    match ExhaustiveUnit {
        ExhaustiveUnit => 0,
        // vvvvv Non-exhaustive variants, should all be ignored
        Unit { .. } => 0,
        Tuple { 0: field, .. } => field,
        StructNoField { .. } => 0,
        Struct { field, .. } => field,
        _ => 0,
    };
}
