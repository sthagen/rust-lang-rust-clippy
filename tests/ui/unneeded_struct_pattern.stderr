error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:18:13
   |
LL |         None {} => 0,
   |             ^^^ help: remove the struct pattern
   |
   = note: `-D clippy::unneeded-struct-pattern` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unneeded_struct_pattern)]`

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:24:13
   |
LL |         None { .. } => 0,
   |             ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:35:18
   |
LL |         Some(None {}) => 0,
   |                  ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:37:13
   |
LL |         None {} => 0,
   |             ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:41:16
   |
LL |     if let None {} = Some(0) {}
   |                ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:43:16
   |
LL |     if let None { .. } = Some(0) {}
   |                ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:45:21
   |
LL |     if let Some(None {}) = Some(Some(0)) {}
   |                     ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:47:13
   |
LL |     let None {} = Some(0) else { panic!() };
   |             ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:49:13
   |
LL |     let None { .. } = Some(0) else { panic!() };
   |             ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:51:18
   |
LL |     let Some(None {}) = Some(Some(0)) else { panic!() };
   |                  ^^^ help: remove the struct pattern

error: this match could be replaced by its body itself
  --> tests/ui/unneeded_struct_pattern.rs:77:5
   |
LL | /     match Custom::Init {
LL | |         _ => 0,
LL | |     };
   | |_____^ help: consider using the match body instead: `0`
   |
   = note: `-D clippy::match-single-binding` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::match_single_binding)]`

error: this match could be replaced by its body itself
  --> tests/ui/unneeded_struct_pattern.rs:82:5
   |
LL | /     match Custom::Init {
LL | |
LL | |         _ => 0,
LL | |     };
   | |_____^ help: consider using the match body instead: `0`

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:96:30
   |
LL |     if let Custom::NoBrackets {} = Custom::Init {
   |                              ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:100:30
   |
LL |     if let Custom::NoBrackets { .. } = Custom::Init {
   |                              ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:104:30
   |
LL |     if let Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} = Custom::Init {
   |                              ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:104:67
   |
LL |     if let Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} = Custom::Init {
   |                                                                   ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:109:43
   |
LL |     if let Custom::NoBracketsNonExhaustive {} = Custom::Init {
   |                                           ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:113:43
   |
LL |     if let Custom::NoBracketsNonExhaustive { .. } = Custom::Init {
   |                                           ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:130:27
   |
LL |     let Custom::NoBrackets { .. } = Custom::Init else {
   |                           ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:134:40
   |
LL |     let Custom::NoBracketsNonExhaustive {} = Custom::Init else {
   |                                        ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:138:40
   |
LL |     let Custom::NoBracketsNonExhaustive { .. } = Custom::Init else {
   |                                        ^^^^^^^ help: remove the struct pattern

error: this match could be replaced by its body itself
  --> tests/ui/unneeded_struct_pattern.rs:157:5
   |
LL | /     match ExhaustiveUnit {
LL | |
LL | |         // Exhaustive variant
LL | |         _ => 0,
LL | |     };
   | |_____^ help: consider using the match body instead: `0`

error: this match could be replaced by its body itself
  --> tests/ui/unneeded_struct_pattern.rs:163:5
   |
LL | /     match ExhaustiveUnit {
LL | |
LL | |         // Exhaustive variant
LL | |         _ => 0,
LL | |     };
   | |_____^ help: consider using the match body instead: `0`

error: aborting due to 23 previous errors

